File: o2a.R

library("httr2")
library("jsonlite")
library("readr")

REGISTRY <- "https://registry.o2a-data.de/rest/v2"
DWS <- "https://dashboard.awi.de/data/rest"

download <- function(url) {
    response <- req_perform(request(url))
    if (response$status_code != 200) {
        stop("Error loading data.")
    } else {
        return(resp_body_json(response))
    }
}

## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
testAggregate <- function(pattern, string) {
    ## :pattern: aggretation pattern to check on
    ## :string: string to be tested
    p <- paste0("[", pattern, "]\\w+")
    a <- grepl(p, string, ignore.case = TRUE)
    return(a)
}

## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
items <- function(pattern = NULL) {
    ## Loads availble sensors from the data service. The optional
    ## pattern allows * wildcards and can be used to search for sensors.
    ## See https://dashboard.awi.de/data/ for documentation.
    ## :pattern: is parameter urn(s)
    url <- paste0(DWS, "/sensors")
    if (!is.null(pattern)) {
        url <- paste0(url, "?pattern=", pattern)
    }
    itemsJson <- download(url)
    return(itemsJson)
}


## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
downloadDataFromDWS <- function(itemUrns,
                                    begin,
                                    end,
                                    aggregate,
                                    aggregateFunctions) {
    ## Loads data from the data service for given sensors
    ## in the given time range and selected aggregate.
    ## See https://dashboard.awi.de/data/ for documentation.
    ## :items: parameter urn(s)
    ## :begin: YYYY-MM-DDTHH:MM:SS string
    ## :end: YYYY-MM-DDTHH:MM:SS string
    ## :aggregate: second, minute, hour, day
    ## :aggregateFunctions: min, max, mean, median, std, count
    if (is.null(itemUrns) || length(itemUrns) == 0) {
        stop("Item(s) must be defined.")
    }
    if (is.null(begin)) {
        stop("Begin timestamp must be defined.")
    }
    if (is.null(end)) {
        stop("End timestamp must be defined.")
    }
    if (grepl(",", itemUrns)) {
        itemUrns <- gsub(" ", "", itemUrns)
        itemUrns <- gsub(",", "&sensors=", itemUrns)
    }
    secondTest <- testAggregate("sec", aggregate)
    minTest <- testAggregate("min", aggregate)
    hourTest <- testAggregate("hour", aggregate)
    dayTest <- testAggregate("day", aggregate)
    ##
    baseLink <- paste0(
        DWS,
        "/data?sensors=",
        itemUrns,
        "&beginDate=",
        as.character(begin),
        "&endDate=",
        as.character(end),
        "&aggregate="
    )
    if (secondTest == TRUE) {
        response <- req_perform(
            request(
                paste0(baseLink,
                "second&streamit=true&withQualityFlags=false",
                "&withLogicalCode=false")
            )
        )
    }
    if (minTest == TRUE) {
        response <- req_perform(
            request(
                paste0(
                baseLink,
                "minute&aggregateFunctions=",
                aggregateFunctions,
                "&streamit=true&withQualityFlags=false",
                "&withLogicalCode=false")
            )
        )
    }
    if (hourTest == TRUE) {
        response <- req_perform(
            request(
                paste0(
                baseLink,
                "hour&aggregateFunctions=",
                aggregateFunctions,
                "&streamit=true&withQualityFlags=false",
                "&withLogicalCode=false")
            )
        )
    }
    if (dayTest == TRUE) {
        response <- req_perform(
            request(
                paste0(
                baseLink,
                "day&aggregateFunctions=",
                aggregateFunctions,
                "&streamit=true&withQualityFlags=false&withLogicalCode=false")
            )
        )
    }
    if (response$status_code != 200) {
        stop(sprintf("Error loading data: %s", response$reason))
    }
    ##
    respBody <- resp_body_string(response)
    df <- read.table(text = respBody, header = TRUE, sep = "\t")
    return(df)
}


## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##

item <- function(code) {
    ## Request and parse item properties for a given item urn as "code"
    ## :code: item unique resource number (urn) or ID
    if (is.character(code)) {
        url <- paste0(REGISTRY, "/items/", code)
        itemJson <- download(url)
    } else if (is.numeric(code)) {
        url <- paste0(REGISTRY, "/items/", code)
        itemJson <- download(url)
    } else {
        stop("provide item urn or item ID")
  }
    ## Adding the items porperties to the Json
    url <- paste0(REGISTRY, "/items/", itemJson$id, "/properties")
    propertiesJson <- download(url)$records
    if (length(propertiesJson) == 0) { #### if no properties
        properties <- NA
    } else {
        # create look up lists
        unitList <- list()
        typeList <- list()
        for (propertyRecord in propertiesJson) {
            propertyUnit <- propertyRecord$unit
            if (length(propertyUnit) > 1) {
                uuid <- propertyUnit$`@uuid`
                unitList[uuid] <- list(propertyUnit)
            }
        }
        for (propertyRecord in propertiesJson) {
            propertyType <- propertyRecord$type
            if (length(propertyType) > 1) {
                uuid <- propertyType$`@uuid`
                typeList[uuid] <- list(propertyType)
            }
        }
        # enhance propertiesJson
        i <- 0
        for (propertyRecord in propertiesJson) {
            i <- i + 1
            propertyUnit <- propertyRecord$unit #uuid
            if (length(propertyUnit) == 1) {
                fullUnit <- unlist(unitList[propertyUnit],
                                    recursive = FALSE,
                                    use.names = FALSE)
                names(fullUnit) <- names(unitList[[1]])
                propertiesJson[[i]]$unit <- fullUnit
            }
        }
        j <- 0
        for (propertyRecord in propertiesJson) {
            j <- j + 1
            propertyType <- propertyRecord$type #uuid
            if (length(propertyType) == 1) {
                fullType <- unlist(typeList[propertyType],
                                    recursive = FALSE,
                                    use.names = FALSE)
                names(fullType) <- names(typeList[[1]])
                propertiesJson[[j]]$type <- fullType
            }
        }
    }
    itemJson$properties <- propertiesJson
    return(itemJson)
}

## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
parameters <- function(code) {
    ## Request....
    ## :code: item ID or urn
    if (is.character(code)) {
        item2 <- item(code)
        code <- item2$id
    } else if (is.numeric(code)) {
        code <- code
        item2 <- item(code)
    } else {
        stop("provide item urn or item ID")
    }
    ##
    url <- paste0(REGISTRY, "/items/", code, "/parameters")
    parameterJson <- download(url)$records
    parameterUrn <- NA
    for (parameterRecord in 1:length(parameterJson)) {
        parameterUrn <- paste0(item2$code, ":",
                                parameterJson[[parameterRecord]]$shortName)
        parameterJson[[parameterRecord]]$urn <- parameterUrn
    }
    return(parameterJson)
}

## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
#code <- 8566
events <- function(code, geo = FALSE) {
    ## Requests all events of an item, returns as dict
    ## :code: registry id of item
    ## :geo: TRUE == only with valid coordinates, FALSE == all events
    if (is.character(code)) {
        code <- item(code)$id
    } else if (is.numeric(code)) {
        code <- code
    } else {
        stop("provide item urn or item ID")
    }
    ##
    if (geo == TRUE) {
        url <- paste0(REGISTRY, "/items/",
                      as.character(code),
                      "/events?where=",
                      "latitude%3E%3D-90%20and%20latitude%3C%3D90%20",
                      "and%20longitude%3E%3D-180%20and%20longitude%3C%3D180"
                      )
    } else {
        url <- paste0(REGISTRY, "/items/", as.character(code), "/events")
    }
    #
    eventsJson <- download(url)$records
    ## create lookup table
    lut <- list()
    for (eventRecord in eventsJson) {
        eventType <- eventRecord$type
        if (length(eventType) > 1) {
            uuid <- eventType$`@uuid`
            lut[uuid] <- list(eventType)
        }
    }
    ## enrich event info
    i <- 0 # variable to count iterations of for loop
    for (eventRecord in eventsJson) {
        i <- i + 1
        if (length(eventRecord$type) == 1) {
            eventType <- eventRecord$type  #event type uuid from record in Json
            fullType <- unlist(lut[eventType],
                                recursive = FALSE,
                                use.names = FALSE)
            names(fullType) <- names(lut[[1]]) # reassign correct names
            eventsJson[[i]]$type <- fullType # paste into correct position in eventsJson
        }
    }
    return(eventsJson)
}



## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##

contacts <- function(code) {
    ## requests contacts per item
    ## :code: registry id or urn of item
    if (is.character(code)) {
        item2 <- item(code)
        code <- item2$id
    } else if (is.numeric(code)) {
        code <- code
    } else {
        stop("provide item urn or item ID")
    }
    ##
    url <- paste0(REGISTRY, "/items/", as.character(code), "/contacts")
    ##
    contactsJson <- download(url)$records
    ## create look up lists
    contactsList <- list()
    rolesList <- list()
    for (contactRecord in contactsJson) {
        contactContact <- contactRecord$contact
        if (length(contactContact) > 1) {
            uuid <- contactContact$`@uuid`
            contactsList[uuid] <- list(contactContact)
        }
        #
        contactRole <- contactRecord$role
        if (length(contactRole) > 1) {
            uuid <- contactRole$`@uuid`
            rolesList[uuid] <- list(contactRole)
        }
    }
    ## enricht contacts info
    i <- 0
    for (contactRecord in contactsJson) {
        i <- i + 1
        if (length(contactRecord$contact) == 1) {
            contactContact <- contactRecord$contact
            fullContact <- unlist(contactContact,
                                recursive = FALSE,
                                use.names = FALSE)
            names(fullContact) <- names(contactsList[[1]])
            contactsJson[[i]]$contact <- fullContact
        }
        #
        if (length(contactRecord$role) == 1) {
            contactRole <- contactRecord$role
            fullRole <- unlist(rolesList[contactRole],
                                recursive = FALSE,
                                use.names = FALSE)
            names(fullRole) <- names(rolesList[[1]])
            contactsJson[[i]]$role <- fullRole
        }
    }
    return(contactsJson)
}

## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
code <- 9215
subitems <- function(code) {
    ## retrieve subitems via code/urn by parent ID
    ## :code: item unique resource number (urn) or ID
    if (is.character(code)) {
        item2 <- item(code)
        code <- item2$id
    } else if (is.numeric(code)) {
        code <- code
    } else {
        stop("provide item urn or item ID")
    }
    ##
    url <- paste0(REGISTRY,
                  "/items?where=parent.id==",
                  as.character(code))
    ##
    subitemsJson <- download(url)$records
    ##
    subitemsTypes <- list()
    for (subitemRecord in subitemsJson) {
        subitemType <- subitemRecord$type
        if (length(subitemType) > 1) {
            uuid <- subitemType$`@uuid`
            subitemsTypes[uuid] <- list(subitemType)
        }
    }
    ## enrich info
    i <- 0
    for (subitemRecord in subitemsJson) {
        i <- i + 1
        subitemType <- subitemRecord$type
        if (length(subitemType) == 1) {
            fullType <- unlist(subitemsTypes[subitemType],
                                recursive = FALSE,
                                use.names = FALSE)
        names(fullType) <- names(subitemsTypes[[1]])
        subitemsJson[[i]]$type <- fullType
        }
    }
    return(subitemsJson)
}


## eof

File: o2a.py
from _io import StringIO
import requests
import re
import pandas as pd
from datetime import date
import json


class o2a:
    """
    This script abstracts access to metadata stored in
    https://registry.o2a-data.de and data available via the data web service
    (https://dashboard.awi.de/data/). Have a look to the documentation at
    https://o2a-data.de/documentation and the API descriptions
    https://registry.o2a-data.de/api/ and https://dashboard.awi.de/data/api/.
    """

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def __init__(self):
        """
        lorem
        """
        self.REGISTRY = "https://registry.o2a-data.de/rest/v2"
        self.DWS = "https://dashboard.awi.de/data/rest"

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def help(self):
        """ """
        print("**DWS**")
        print(
            "- x.items: retrieves info about a specific item if 'code'\n (item urn) is submitted otherwise it gives a de facto unmanagable\n data stream to std out. Info is derived from the data web service."
        )
        print(
            "- x.downloadDataFromDWS: retrieves data from data web service \ninput: i) string of parameter urns, multiple urns can be combined by comma, ii) beginning date (YYYY-MM-DDTHH:MM:SS), iii) end date (YYYY-MM-DDTHH:MM:SS), iv) aggregate level (second, minute, hour, day), defaults to 'hour', iv) aggregate function (min, max, mean, median, count, std), does not apply for 'seconds', defaults to mean"
        )
        print("")
        print("**REGISTRY**")
        print(
            "- x.item: full json of item (incl. item properties)\ninput: item ID or urn"
        )
        print("- x.contacts: full json of item contacts\ninput: item ID or urn")
        print(
            "- x.events: full json list of events \ninput: item ID or urn, optional 'geo = True' only events with coordinates are put out"
        )
        print("- x.subitems: output of all subitems \ninput: item ID or urn")
        print("- x.parameters: gives all parameters per item \ninput: item ID or urn")
        print(
            "- x.resources: return list of available resources\ninput: item ID or urn"
        )

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def _download(self, url):
        """
        auxiliary function
        :url: externally created string to call by this fun
        """
        response = requests.get(url)
        if response.status_code != 200:
            raise Exception("Error loading data.".format(response.reason))
        else:
            return json.loads(response.content)

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def _testAggregate(self, pattern, string):
        """
        :pattern: aggretation pattern to check on
        :string: string to be tested
        """
        p = "[" + pattern + "]\\" + "w+"
        a = re.match(p, string.lower())
        if a is None:
            return False
        else:
            return True

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def items(self, pattern: str = None):
        """
        Loads availble sensors from the data service. The optional
        pattern allows * wildcards and can be used to search for sensors.
        See https://dashboard.awi.de/data/ for documentation.
        :pattern: is parameter urn(s)
        """
        url = self.DWS + "/sensors"
        if pattern != None:
            url += "?pattern=" + pattern

        j = self._download(url)
        return j

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def downloadDataFromDWS(
        self,
        items,
        begin: date,
        end: date,
        aggregate: str = "hour",
        aggregateFunctions: str = "mean",
    ):
        """
        Loads data from the data service for given sensors
        in the given time range and selected aggregate.
        See https://dashboard.awi.de/data/ for documentation.
        :items: parameter urn(s)
        :begin: YYYY-MM-DDTHH:MM:SS string
        :end: YYYY-MM-DDTHH:MM:SS string
        :aggregate: second, minute, hour, day
        :aggregateFunctions: min, max, mean, median, std, count, ...
        """
        if items == None or len(items) == 0:
            raise Exception("Item(s) must be defined.")

        if begin == None:
            raise Exception("Begin timestamp must be defined.")

        if end == None:
            raise Exception("End timestamp must be defined.")

        if items.count(",") > 0:
            items = items.replace(" ", "").replace(",", "&sensors=")

        secondTest = self._testAggregate("sec", aggregate)
        minTest = self._testAggregate("min", aggregate)
        hourTest = self._testAggregate("hour", aggregate)
        dayTest = self._testAggregate("day", aggregate)

        baseLink = (
            self.DWS
            + "/data?sensors="
            + items
            + "&beginDate="
            + str(begin)
            + "&endDate="
            + str(end)
            + "&aggregate="
        )

        if secondTest is True:
            response = requests.get(
                baseLink
                + "second&streamit=true&"
                + "withQualityFlags=false&withLogicalCode=false"
            )

        if minTest is True:
            response = requests.get(
                baseLink
                + "minute&aggregateFunctions="
                + aggregateFunctions
                + "&streamit=true&withQualityFlags=false&withLogicalCode=false"
            )

        if hourTest is True:
            response = requests.get(
                baseLink
                + "hour&aggregateFunctions="
                + aggregateFunctions
                + "&streamit=true&withQualityFlags=false&withLogicalCode=false"
            )

        if dayTest is True:
            response = requests.get(
                baseLink
                + "day&aggregateFunctions="
                + aggregateFunctions
                + "&streamit=true&withQualityFlags=false&withLogicalCode=false"
            )

        ##+ '&limit=20' # <-------------------------------------------- ??!?
        if response.status_code != 200:
            raise Exception("Error loading data.".format(response.reason))

        # build the data frame
        df = pd.read_csv(StringIO(response.text), sep="\t")
        df["datetime"] = pd.to_datetime(df["datetime"])
        return df

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def item(self, code):
        """
        Request and parse item properties for a given item urn as "code"
        :code: item unique resource number (urn) or ID
        """
        ##        if sys == "dev": ## later

        if type(code) == str:
            url = self.REGISTRY + "/items?where=code=LIKE=" + code
            j = self._download(url)["records"][0]
        elif type(code) == int:
            url = self.REGISTRY + "/items/" + str(code)
            j = self._download(url)
        else:
            raise Exception("provide item urn or item ID")

        ## ITEM properties
        url = self.REGISTRY + "/items/" + str(j["id"]) + "/properties"
        k = self._download(url)["records"]
        j["itemProperties"] = k

        return j

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def parameters(self, code):
        """
        Request....
        :code: item ID or urn
        """
        item = self.item(code)
        if type(code) == str:
            code = item["id"]
        elif type(code) == int:
            code = code
        else:
            raise Exception("provide item urn or item ID")

        url = self.REGISTRY + "/items/" + str(code) + "/parameters"
        k = self._download(url)["records"]
        for i in range(len(k)):
            k[i]["urn"] = item["code"] + ":" + k[i]["shortName"]
        ##
        return k

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def events(self, code, geo=False):
        """
        Requests all events of an item, returns as dict
        :code: registry id of item
        :geo: true == only with valid coordinates, false == all events
        """
        if type(code) == str:
            item = self.item(code)
            code = item["id"]
        elif type(code) == int:
            code = code
        else:
            raise Exception("provide item urn or item ID")

        if geo is True:
            url = (
                self.REGISTRY
                + "/items/"
                + str(code)
                + "/events?where=latitude>=-90 and latitude<=90 and"
                + "longitude>=-180 and longitude<=180"
            )
        else:
            url = self.REGISTRY + "/items/" + str(code) + "/events"

        j = self._download(url)["records"]

        ## create lookup table
        lut = {}
        for i in j:
            if isinstance(i["type"], dict):
                lut[i["type"]["@uuid"]] = i["type"]

        ## enrich event info
        for i in range(len(j)):
            if not isinstance(j[i]["type"], dict):
                j[i]["type"] = lut.get(j[i]["type"])

        return j

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def contacts(self, code):
        """
        requests contacts per item
        :code: registry id or urn of item
        """
        if type(code) == str:
            item = self.item(code)
            code = item["id"]
        elif type(code) == int:
            code = code
        else:
            raise Exception("provide item urn or item ID")

        url = self.REGISTRY + "/items/" + str(code) + "/contacts"

        j = self._download(url)["records"]

        contacts, roles = {}, {}
        for i in j:
            if isinstance(i["contact"], dict):
                contacts[i["contact"]["@uuid"]] = i["contact"]
            if isinstance(i["role"], dict):
                roles[i["role"]["@uuid"]] = i["role"]

        for i in range(len(j)):
            if not isinstance(j[i]["contact"], dict):
                j[i]["contact"] = contacts.get(j[i]["contact"])
            if not isinstance(j[i]["role"], dict):
                j[i]["role"] = roles.get(j[i]["role"])

        return j

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##
    def subitems(self, code):
        """
        retrieve subitems via base url and "*"
        :code: item unique resource number (urn) or ID
        """
        if type(code) == str:
            item = self.item(code)
            code = item["id"]
        elif type(code) == int:
            code = code
        else:
            raise Exception("provide item urn or item ID")

        url = self.REGISTRY + "/items?where=parent.id==" + str(code)
        j = self._download(url)["records"]

        types = {}
        for i in j:
            if isinstance(i["type"], dict):
                types[i["type"]["@uuid"]] = i["type"]

        for i in range(len(j)):
            if not isinstance(j[i]["type"], dict):
                j[i]["type"] = types.get(j[i]["type"])
        return j

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##

    def resources(self, code):
        """
        retrieve resources of item
        :code: item unique resource number (urn) or ID
        """
        if type(code) == str:
            url = self.REGISTRY + "/items?where=code=LIKE=" + code
            j = self._download(url)["records"][0]
        elif type(code) == int:
            url = self.REGISTRY + "/items/" + str(code)
            j = self._download(url)
        else:
            raise Exception("provide item urn or item ID")

        ## ITEM resources
        url = self.REGISTRY + "/items/" + str(j["id"]) + "/resources"
        k = self._download(url)["records"]

        resourceIds = [i["id"] for i in k]
        ##
        c = []
        for i in resourceIds:
            url = self.REGISTRY + "/items/" + str(j["id"]) + "/resources/" + str(i)
            l = self._download(url)
            d = {}
            d["name"] = l["name"]
            d["type"] = l["type"]["generalName"]
            d["description"] = l["description"] if "description" in l.keys() else ""
            d["link"] = l["linkage"] if "linkage" in l.keys() else url + "/payload"

            c.append(d)

        return c

    ## ---------------------------  ¬!"£$%^&*()_+ --------------------------- ##


## eof
